tensorboard/loader.py

```diff
@@ -732,85 +740,367 @@ def get_next_event(self):
 
     :rtype: tf.Event
     """
+    record = self._reader.get_next_record()
+    if record is None:
+      return None
+    event = tf.Event()
+    event.ParseFromString(record.record)
+    self._offset = record.offset
+    self._update_first_step(event)
+    return event
+
+  def set_offset(self, offset):
+    """Sets byte offset in file.
+
+    :type offset: int
+    """
+    if offset == self._offset:
+      return
+    self._reader.close()
+    self._reader = self._reader_factory(self.path, offset)
+    self._offset = offset
+
+  def get_offset(self):
+    """Returns current byte offset in file.
+
+    :rtype: int
+    """
+    return self._offset
+
+  def get_size(self):
+    """Returns byte length of file.
+
+    :rtype: int
+    """
+    return self._reader.get_size()
+
+  def get_first_step(self):
+    """Determines step count of first summary event.
+
+    This method requires an open/read/close but memoizes the result. If
+    the answer was already inferred from previous calls to
+    get_next_event, then no i/o should be necessary.
+
+    Returns:
+      The step count of the first summary event, or None if there
+      didn't appear to be any summary events, or the user isn't logging
+      step counts.
+
+    :rtype: int
+    """
+    if self._first_step is not None:
+      return self._first_step
+    if self._update_first_step_invocations >= EventLogReader.FIRST_STEP_PEEKS:
+      return None
+    with RecordReader(self.path) as reader:
+      for _ in range(EventLogReader.FIRST_STEP_PEEKS):
+        record = reader.get_next_record()
+        if record is None:
+          break
+        event = tf.Event()
+        event.ParseFromString(record.record)
+        self._update_first_step(event)
+    return self._first_step
+
+  def close(self):
+    """Closes event log reader if open.
+
+    Further i/o is not permitted after this method is called.
+    """
+    if self._reader is not None:
+      self._reader.close()
+      self._reader = None
+
+  def _update_first_step(self, event):
+    self._update_first_step_invocations += 1
+    # proto3 doesn't let us distinguish between absent and 0.
+    if event.step:
+      step = 0 if self._observed_zero_step else event.step
+      if self._first_step is None or step < self._first_step:
+        self._first_step = step
+    elif event.summary:
```

Comment:
I thought about it some more. It might be better if we just leave this optimization out for now.